### PR TITLE
hoist commas out of memcmp expressions

### PR DESCRIPTION
**Pull request by WalterBright**
These sorts of rewrites improve optimization by reducing the variety of patterns that need to be recognized. It also makes the ASTs easier to read.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/backend/cgelem.d [hunk 1]
```diff
@@ -5746,4 +5784,22 @@ private elem *elToPair(elem *e)
     return e;
 }
 
+/******************************************
+ * Determine if `b` can be moved before `a` without disturbing
+ * order-of-evaluation semantics.
+ */
+
+private bool canHappenAfter(elem* a, elem* b)
```
Comment (jacob-carlborg) [inline src/dmd/backend/cgelem.d]: Would `canMoveAfter` be a better name?

**Comment by jacob-carlborg** [comment]
Testable?

File: src/dmd/backend/cgelem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/backend/cgelem.d]: Perhaps, but the uses of it actually move `b` before `a`.

**Comment by WalterBright** [comment]
It not breaking things is already covered by existing tests.

File: src/dmd/backend/cgelem.d (hunk 1 above)
Comment (jacob-carlborg) [reply src/dmd/backend/cgelem.d]: So, `canMoveBefore` or `canPrepend`? `canHappenAfter` seems like a very arbitrary name.

**Pull request merged**